### PR TITLE
improve release handling 

### PR DESCRIPTION
**Pull request by sennui**

### CODE REVIEW

**Comment by zack-bitcoin** [comment]
I can't get it to turn on.

```
zack:~/Hacking/sennui_testnet$ make compile
===> Verifying dependencies...
===> Compiling ae_api
===> Compiling ae_core
zack:~/Hacking/sennui_testnet$ make local-release-start
make: ./_build/local/rel/ae_core/bin/ae_core: Command not found
Makefile:34: recipe for target 'local-release-start' failed
make: *** [local-release-start] Error 127
```

**Comment by sennui** [comment]
added missing step in README: `make local-release`

**Comment by zack-bitcoin** [comment]
Where is the keys.db file?
I want to replace it with the master private key so I can run tests.

**Comment by sennui** [comment]
its in _build//local/rel/ae_core/data/keys.db

**Comment by sennui** [comment]
can you, for now, do the make local-release-attach and add the peer from console - I will add it to config file

**Comment by sennui** [comment]
added

**Comment by zack-bitcoin** [comment]
I did a git pull, then ran `make local-release`, then turned it on and attached to it.
peers:all() returns an empty list.

What is the correct way to rebuild when the code changes?

peers:add() works to add the server.

**Comment by zack-bitcoin** [comment]
How do I do the equivalent of clean.sh
I need to delete the blocks in order to run tests.

**Comment by sennui** [comment]
We were not aware about this use case for clean.sh

So, for now just run clean.sh (its in scripts dir)

To make it work you need to prefix all path with location of the node:
e.g.:

`../_build/local/rel/ae_core/data/keys.db`

instead of old:

`data/keys.db`

**Comment by sennui** [comment]
@zp-sd please add this flow to Makefile, so it also covers 3-node deployment (it can live in Makefile only or call a script)

**Comment by zack-bitcoin** [comment]
who is zp-sd?

cleaning didn't work.

```
zack:~/Hacking/sennui_testnet/scripts$ sh clean.sh 
clean.sh: 3: clean.sh: ../_build/local/rel/ae_core/data/keys.db: Permission denied
rm: cannot remove 'data': No such file or directory
rm: cannot remove 'blocks': No such file or directory
zack:~/Hacking/sennui_testnet/scripts$ sudo sh clean.sh 
[sudo] password for zack:
clean.sh: 3: clean.sh: ../_build/local/rel/ae_core/data/keys.db: Permission denied
```

**Comment by michalzajda** [comment]
Zp-sd is team member from sennui

**Comment by zack-bitcoin** [comment]
It is important when I do tests with 3 nodes that clean.sh works, because I want to do lots of tests and be able to start from the same state each time.

**Comment by sennui** [comment]
@zp-sd is team member from sennui.

**Comment by sennui** [comment]
@zack-bitcoin 

something like that:

```
if [ -e "../_build/local/rel/ae_core/data/keys.db" ]
then
    cp ../_build/local/rel/ae_core/data/keys.db ../_build/local/rel/ae_core/keys_backup
fi
echo `pwd`
rm -r ../_build/local/rel/ae_core/data
mkdir ../_build/local/rel/ae_core/data
rm -r ../_build/local/rel/ae_core/blocks
mkdir ../_build/local/rel/ae_core/blocks

if [ -d "../_build/local/rel/ae_core/backup" ]
then
    touch ../_build/local/rel/ae_core/backup/temp.db
    rm ../_build/local/rel/ae_core/backup/*.db
else
    mkdir ../_build/local/rel/ae_core/backup
fi

#if [ -e "../_build/local/rel/ae_core/data/keys_backup" ]
#then
cp ../_build/local/rel/ae_core/keys_backup ../_build/local/rel/ae_core/data/keys.db
#fi

```

**Comment by sennui** [comment]
@zack-bitcoin btw so this is next thing to fix: if key.db is supposed to survive between dev block cleanups it should have its own dir

**Comment by sennui** [comment]
@zack-bitcoin exactly, you are right. With 3 nodes its a lot of mess. We were not aware about master key meaning. We will fix it

**Comment by zack-bitcoin** [comment]
yes, moving keys.db to a different directory sounds like a good idea.

**Comment by zp-sd** [comment]
@zack-bitcoin @sennui The following changes are prepared:
1) `key.db` file is moved to `keys/keys.db` (and `keys` directory is not affected by cleanup targets)
2) cleanup commands are added to purge appropriate environment (they clean all data in `blocks` and `data` dirs):
* `make local-release-clean` - clean local env
* `make clean-dev1` - clean only `dev1` env
* `make clean-dev2` - clean only `dev2` env
* `make clean-dev3` - clean only `dev3` env
* `make dev-release-clean` - clear all 3 dev envs

The changes are here - https://github.com/sennui/testnet/pull/2.

However @sennui forgot to grant me write/merge permissions to `sennui/testnet` repo, and we have to wait for him to add these changes to this PR.

**Comment by sennui** [comment]
clean makefile targets are in

**Comment by zack-bitcoin** [comment]
local-release-clean works. The single-node tests all pass.
Now I will try to run the multi-node tests.

**Comment by zack-bitcoin** [comment]
There is a problem with starting 3 nodes in dev mode.
Before, when I ran `sh setup.sh` it created 3 copies of the erlang directory instantly, and I was ready to run the test. setup.sh did not run the erlang compiler, the pre-compiled code was copied.

Now when I run `make dev-release` it seems to be building 3 different releases.

It is important that I am able to update the code on all these nodes quickly, because I want to be able to run the integration test in as little time as possible after changing the code.

**Comment by sennui** [comment]
@zack-bitcoin yes, the whole point of these kind of changes is to improve the flow of development.

There is `default` dir in `_build` dir. From over there the code is copied/linked to more locations. 
Its not necessarly compiled 3 times - we will check that

Proper OTP release is daemonized, so re-loading code may have a bit bigger overlay.

Make change -> make release -> kill current nodes -> start nodes

Hot loading code is dangerous in case of stateful processes (like gen_srv, gen_fsm)

Side note: currently Ctrl+C in attached session killes the node. I think we should change it and Ctrl+C will just detach from the deamon. Only make stop will kill it

The project is really small so maybe just yet another make target to wrap those above will be fine?

@zp-sd can you check that? (I am on vacation:)

**Comment by sennui** [comment]
@zack-bitcoin I can build Emakefile that will hot load the code to the current VM (and recompile it before loading) but it will not restart gen_srv. Maybe that will be handy anyways?

**Comment by zack-bitcoin** [comment]
I need to run tests with at least 3 nodes. Each node needs to have it's own database of blocks, they need to each have their own gen_server for verifying blocks.
I need to be able to turn on 3 nodes almost instantly.

I suspect that the setup.sh I had written before could still be used for this purpose.

**Comment by sennui** [comment]
@zack-bitcoin now when you run `make dev-release` you get three diffent sub trees in `_build` dir 

Each of them has its own db, keys, blocks etc. Completely isolated.

Each of them has also a config with stuff that you used to put in the dev(1-3)_mode shell script (keys, passes etc.) You can adjust them in `config/dev{1-3}/sys.config`

**Comment by zack-bitcoin** [comment]
The test for sharing blocks between two nodes works well.
The test for doing a lightning payment with 3 nodes is failing.

**Comment by sennui** [comment]
@zack-bitcoin can we get step by step execution that is failing? (can you just drop smth like bash history + configs to see if we actually run them according to your design?)

**Comment by zack-bitcoin** [comment]
Having me send you errors is a bad way to do development.
I had to start all three dev nodes in different terminals to be able to see the errors. I should be outputting everything to log files, but I haven't done that yet.

To run the tests, you need to make a couple small changes.
free_constants:test_mode() needs to be set to true.
in the config file, master_pubkey needs to be the same as the pubkey stored in your wallet. you can find your pubkey like this `keys:pubkey()`
the id stored in your wallet needs to be a 1. you can look up your id like this `keys:id()`. you can change your id like this `keys:update_id(1)`

**Comment by zack-bitcoin** [comment]
I ran the lightning test one more time with the aeternity testnet code. It still passes.

**Comment by zp-sd** [comment]
@zack-bitcoin Ok, thank you for your response regarding test setup.

Test `share_blocks_test.sh` is working on our env as well and we were able to reproduce failures in `lightning_tests.sh`. So I believe we're in the same place with our local setup now and we're working on these failures now.

**Comment by sennui** [comment]
@zack-bitcoin to be honest its one of the very few ways to debug bugs. Given that our environments are different (because we don't operate on dedicated host/vm/container) We will need git diff and logs (you get them for free in _build's subdir)

We are rebasing again and trying to pinpoint change in behaviour between main repo and the PR

**Comment by sennui** [comment]
@zack-bitcoin so we got new set of fixes and the same stuff fails for me in master and in branch...
so there is good chance we fixed stuff. 

Take a look how I changed master key handling:
1. A key is committed to tests directory and its used when making 3-node deployment
2. Master pubkey is extracted to configs (from constans.erl)
3. Pubkey of the key from #1 is committed to configs (as master key)
To sum up, no need to manipulate code/keys for testing 

I can do the same for one node deployment, but I need to talk to you if its useful

**Comment by sennui** [comment]
This is the way I set up 3 node deployment:

1. kill all running nodes
2. `make dev-release-clean`
3. `make dev-release-unikey`
4. `make dev-release-start` 

Nodes are setup with test-key and master key is pubkey of the test-key

**Comment by sennui** [comment]
@zack-bitcoin added extra configs for local release in config dir. This way when you build one node local target it is in test mode true. The prod release is test mode false (also one node)

resolved conflicts

**Comment by sennui** [comment]
This https://github.com/BumblebeeBat/pink_crypto/pull/1 PR is blocking from fixes here
